mdtraj/geometry/src/geometry.cpp

```diff
@@ -0,0 +1,313 @@
+/*=======================================================================*/
+/* MDTraj: A Python Library for Loading, Saving, and Manipulating        */
+/*         Molecular Dynamics Trajectories.                              */
+/* Copyright 2012-2016 Stanford University and the Authors               */
+/*                                                                       */
+/* Authors: Peter Eastman                                                */
+/* Contributors:                                                         */
+/*                                                                       */
+/* MDTraj is free software: you can redistribute it and/or modify        */
+/* it under the terms of the GNU Lesser General Public License as        */
+/* published by the Free Software Foundation, either version 2.1         */
+/* of the License, or (at your option) any later version.                */
+/*                                                                       */
+/* This library is distributed in the hope that it will be useful,       */
+/* but WITHOUT ANY WARRANTY; without even the implied warranty of        */
+/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         */
+/* GNU Lesser General Public License for more details.                   */
+/*                                                                       */
+/* You should have received a copy of the GNU Lesser General Public      */
+/* License along with MDTraj. If not, see <http://www.gnu.org/licenses/>.*/
+/*=======================================================================*/
+
+#include "geometry.h"
+#include "vectorize_sse.h"
+#include <cmath>
+#include <cstddef>
+#include <float.h>
+#include <vector>
+
+using std::vector;
+
+/****************************************************************************/
+/* Distance, Angle and Dihedral kernels                                     */
+/****************************************************************************/
+
+/**
+ * This is kindof hacky / gross, but I think it's the best way to avoid having
+ * lots of copy-paste code. For each of the distance/angle/dihedral kernels, we
+ * want to compile two version: one which uses PBCs and the other which does
+ * not. Most of the code between these versions is shared, so I've written
+ * the parts of the two functions which are different using #ifdefs. So we
+ * just include these files _twice_ here, toggling the variable that controls
+ * the ifdef.
+ *
+ * Note that these kernel files are really not capable of being compiled
+ * independently -- they're not header files at all -- and they're really just
+ * meant to be #included here.
+ **/
+#undef COMPILE_WITH_PERIODIC_BOUNDARY_CONDITIONS
```

Comment:
> (Personally, I don't find it all that hacky, but whatever.)

I remember thinking exactly the same thing when I saw it in `geometry.c` :smiley:.  This is a fairly common technique I've seen used in many places including several codes in Amber as well as OpenMM (in a sense, anyway).